jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,297 @@
+                int read = _byteChannel.read(byteBuffer);
+                BufferUtil.flipToFlush(byteBuffer, 0);
+                if (read == 0)
+                    return null;
+                if (read > 0)
+                {
+                    _totalRead += read;
+                    _buffer.retain();
+                    if (_length < 0 || _totalRead < _length)
+                        return Content.Chunk.asChunk(byteBuffer, false, _buffer);
+
+                    Content.Chunk last = Content.Chunk.asChunk(byteBuffer, true, _buffer);
+                    lockedSetTerminal(Content.Chunk.EOF);
+                    return last;
+                }
+                lockedSetTerminal(Content.Chunk.EOF);
+            }
+            catch (Throwable t)
+            {
+                lockedSetTerminal(Content.Chunk.from(t, true));
+            }
+        }
+        return _terminal;
+    }
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            lockedSetTerminal(Content.Chunk.from(failure, true));
+        }
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public boolean rewind()
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            // We can remove terminal condition for a rewind that is likely to occur
+            if (_terminal != null && !Content.Chunk.isFailure(_terminal) && (_byteChannel == null || _byteChannel instanceof SeekableByteChannel))
+                _terminal = null;
+
+            lockedEnsureOpenOrTerminal();
+            if (_terminal != null || _byteChannel == null || !_byteChannel.isOpen())
+                return false;
+
+            if (_offset >= 0 && _byteChannel instanceof SeekableByteChannel seekableByteChannel)
+            {
+                try
+                {
+                    seekableByteChannel.position(_offset);
+                    _totalRead = 0;
+                    return true;
+                }
+                catch (Throwable t)
+                {
+                    lockedSetTerminal(Content.Chunk.from(t, true));
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * A {@link ByteChannelContentSource} for a {@link Path}
+     * @deprecated To be replaced by an updated {@link org.eclipse.jetty.io.content.PathContentSource} in 12.1.0
+     */
+    @Deprecated(forRemoval = true, since = "12.0.11")
+    public static class PathContentSource extends ByteChannelContentSource
+    {
+        private final Path _path;
+
+        public PathContentSource(Path path)
+        {
+            super(new ByteBufferPool.Sized(null), null, 0, size(path));
+            _path = path;
+        }
+
+        public PathContentSource(ByteBufferPool.Sized byteBufferPool, Path path)
+        {
+            super(byteBufferPool, null, 0, size(path));
```

Comment:
Not sure the `size(path)` is necessary, can't we just do `length=-1` and read to EOF and save the extra fileSystem access to look up the file size?

Reply:
The previous impl did that, but I don't see us saving much, other than for tiny files.  So OK.

Reply:
Turns out that it is needed by the multipart tests that expect to known how long a part is.  reverting the change.